### ISSUE TITLE
TDI mode cannot be set to True

### ISSUE THREAD

**Issue by taishan1** [EduardoVB/NewTab#4]
Steps to reproduce:
1. Start a new project
2. Add NewTab for VB6 in components dialog
3. Add NewTab to default form
4. Accept default name of "NewTab", or change to anything else
5. Try to set property TDIMode to True
6. Message: This property needs to be set inmediately (sic) after adding the NewTab control (without changing other properties first)

**Comment by EduardoVB** [comment]
Cannot reproduce. Are you using a previous version or the current one?

**Comment by taishan1** [comment]
November 5 edition .ocx

**Comment by taishan1** [comment]
I have created custom themes with previous versions on other projects. Is there a chance of registry residue?

**Comment by EduardoVB** [comment]
Check if the key HKEY_CURRENT_USER\Software\VB and VBA Program Settings\NewTab exists, if it exists, please export its content and paste it here.

**Comment by taishan1** [comment]
No registry key exists.
TDI documentation is scarce. 
Only reason I wanted to try TDI is because of poor performance changing tabs while using an anchor.ocx. 
Will TDI improve performance?

**Comment by EduardoVB** [comment]
I'm unable to reproduce the issue because it does not happen here with the same ocx version (that is the last/current one).

If there are no registry keys, then something must be changing some property.
Maybe an add-in?

One thing you could do is to use the NewTab in source code, use it in a test project, put a break point at the line `mControlJustAdded = False` and at `mControlJustAdded = nValue` too.
Also at the two lines: `MsgBox "This property needs to be set i...`

And when the program stops in one of then, press Control+L and check from where it is being called (and post the screenshot here).

> Only reason I wanted to try TDI is because of poor performance changing tabs while using an anchor.ocx.
Will TDI improve performance?

I don't understand the question. TDI is for Tabbed Documents Interface, as opposed to MDI (Multiple Documents Interface) and SDI (Single Document Interface).

It does not have to do with "performance". What performance are you talking about?
Anyway, a tab control is unlikely to have anything to do with "performance", it is only a visual control.

**Comment by taishan1** [comment]
Thank you for the source code suggestion.
In regards to my question about improving performance, there is a delay when switching tabs and it seems to be a moving/repainting controls issue.
Without stepping through the source code, this appears to be what is slowing performance:
1. Click a new tab, the ocx moves the contained controls from offscreen left or right
2. The anchor control resizes the controls, and paints them
3. The tab control then repaints the controls again
I've included the newtab.refresh event and it seems to marginally help. I will test the .ocx as a .ctl to find out more information. In the meantime, it would be helpful to see screenshots of the difference between newtab TDI, MDI, and SDI so I can visualize if we are on the same page.

**Comment by EduardoVB** [comment]
I have no idea about that "anchor control" and how it works together with the NewTab control, but yes, there is a small delay when changing tabs due to a visual effect.

To change that, set the `TabTransition` property to `ntTransitionImmediate` (and please tell me if that solved your issue)

**Comment by taishan1** [comment]
Yes, I have already tried that setting. I will just live with the performance as it is. Thanks for an excellent tab control.

**Comment by EduardoVB** [comment]
If you could make a sample project where you can show the "performance" issue I could see it.
There is no much performance involved, just to show and hide controls.
My guess is that the issue must be some other control, that anchor control or something else.

**Comment by taishan1** [comment]
Please recommend an anchor / resize / move controls solution that is compatible with NewTab01... obviously, a user will resize the form, so what is the best way to resize a form's control with NewTab01? Thanks

**Comment by EduardoVB** [comment]
I can't recommend because I don't use such controls, I do that by code.
I don't think that it is a good idea to automatically resize controls in most of cases.

**Comment by taishan1** [comment]
I start the form at a common laptop resolution of 1366x768. If the user maximize the form on 1920x1080, it looks strange for the listviews to not size with the form.

Also, I am not sure if it is just my system, but when the Windows graphics Scale is set to anything but 100%, for example 200%, the version 4.0 of newtab01 changes scale in a weird way in the IDE...?

**Comment by EduardoVB** [comment]
Add exceptions to the resizer for not to handle these controls, and position/size the tab control and its contained controls by code.

There is another option: put the content of each tab into a PictureBox containing all the controls of that tab and see if that fixes it.

> when the Windows graphics Scale is set to anything but 100%, for example 200%, the version 4.0 of newtab01 changes scale in a weird way in the IDE...?

The tab control sets the non-visible controls (the controls in tabs that are not the active one) to a left position shifted, for example -75000 twips. This is the same as the SSTab control does.

But... the problem is that at 200% DPI or 300% DPI the screen does not have 75000 twips to the left, but less. 
Then instead of shifting -75000 twips, I shift less, based on the DPI setting.
And if the DPI setting changes, all is converted automatically.

But... if a resizer control stores some values at some DPI setting, and the DPI setting changes... yes, those stored values won't work correctly.

For sizing/placing controls I made an event: `TabBodyResize`

for example:

```
Private Sub NewTab1_TabBodyResize()
    NewTab1.ControlMove "Picture1", NewTab1.TabBodyLeft, NewTab1.TabBodyTop, NewTab1.TabBodyWidth, NewTab1.TabBodyHeight
End Sub
```

For getting or setting the left of the controls, there is the method `.ControlLeft` and for moving the controls `.ControlMove`

Why? Because if the controls are not in the active tab, their Left property that you can read is changed, and if you set it, they will go to a wrong position. So use `NewTab1.ControlLeft` and `NewTab1.ControlMove` instead.
With the SSTab control there is the same issue but it does not have these helper functions.

So, in summary, try with pictureboxes on each tab containing all the controls. If that does not work, then place/size the controls by code in the TabBodyResize event using `NewTab1.ControlLeft` and `NewTab1.ControlMove` if you are going to work with controls that are in tabs that are not the active one (sizing positioning non currently visible controls).

I think I'll have to write a help for this, but it is basically what I tried to explain here.

A third option is to use another tab control, a tabstrip. I guess the resizer should work with the tabstrip control.

**Comment by EduardoVB** [comment]
PS: I just issued an update with several minor bug fixes.

**Comment by taishan1** [comment]
Thank you Eduardo! You're the best. I will install the update, and I will try your suggestions and report back ASAP.

**Comment by taishan1** [comment]
Eduardo, your suggestion of using 'picturebox as a container for controls' works with Scale <> 100%, and with the resizing anchor control as well. It also seems to switch tabs more efficiently with this new update. Thanks again for everything!

EDIT: After restarting Windows with 200% scale, my IDE problem came back. The tab control is resizing itself in a loop, which may be subclassing issue?

**Comment by EduardoVB** [comment]
Hello, I would need a sample project and/or the steps to reproduce the issue.

**Comment by taishan1** [comment]
the following is repeatable under Windows 10 & 11, without anchor controls

1. right-click desktop, select Display Settings
2. under Scale & Layout, select 100% scale (all monitors) 
3. start new .exe project in vb6, add VBCCR17 and newTab components
4. add form and set scalemode to pixels 
5. size form to 1366 x 768
6. Add tab control, make 4 tabs, segoe mdl2 assets icons for each, and a caption
7. add a VBCCR17 listview and listboxW to each tab, run and save.
8. go back to Display settings and set windows scale to 200% (all monitors)
9. restart Windows, open this vb6 project

**Comment by EduardoVB** [comment]
Hello, I could not replicate the issue, it does not happen here.
Please make a sample project, zip it and upload it here, so we can be sure we are testing the same thing (and verify that the issue happens in that sample project).

**Comment by EduardoVB** [comment]
Please test now, I made a change that may have fixed the issue.

**Comment by taishan1** [comment]
Excellent. You have definitely improved the issue in the IDE. However, some strange artifacts remain. 
EDIT: When you load the project in a different scale than 100%, the tabs disappear and the Segoe MDL icons get stacked on top of each other on the far left. 
You can fix this by selecting a theme and applying it in the customize tab dialog.
This still leaves residual sizing issues in the IDE and EXE, but I believe it is able to be 'worked around'
So, I will say the issue is 80% fixed. The tab control is now editable in the IDE where before you couldn't modify anything. So, this is definitely moving in the right direction.
It might be 100% fixed, but there are so many controls (including my picBox containers for listView and ListBoxW) and other variables that I haven't had time to test all of the permutations. I will keep you informed as I progress.
Thanks Eduardo!

**Comment by EduardoVB** [comment]
It must be something with your tab: my guess: many tabs and little screen space.
But anyway I don't know how to recreate it, so I need a sample project where the issue occurs.

**Comment by EduardoVB** [comment]
Hello, I made another change that might improve the issue at run-time. The project is updated.

**Comment by taishan1** [comment]
Hi Eduardo, I had to revert to the previous version. 
The newest version at runtime and compiled .exe makes the tabs disappear -
and the Segoe MDL icons get stacked on top of each other on the far left. 
BTW: The tabs have plenty of room to display icons and captions at the current size.

**Comment by EduardoVB** [comment]
I need a sample project (simple, basic) to be able to reproduce the issue, and then study what is happening.

**Comment by taishan1** [comment]
I sent an email to Balkesoft over a week ago and never received a reply.

**Comment by EduardoVB** [comment]
It was in the spam folder.

**Comment by EduardoVB** [comment]
I answered the email, I don't know if you received it.

Anyway I issued an update. I'm quite sure that this version should not have the issue you mentioned regarding the previous version.
